File: Pandacap.LowLevel/Data/TheirContent/DeviantArt/DeviantArtFavorite.fs
namespace Pandacap.Data

open System
open Pandacap.PlatformBadges

type DeviantArtFavorite() =
    member val Id = Guid.Empty with get, set
    member val CreatedBy = Guid.Empty with get, set
    member val Username = "" with get, set
    member val Usericon = nullString with get, set
    member val Timestamp = DateTimeOffset.MinValue with get, set
    member val Title = nullString with get, set
    member val Content = nullString with get, set
    member val LinkUrl = nullString with get, set

    member val ThumbnailUrls = new ResizeArray<string>() with get, set

    member val FavoritedAt = DateTimeOffset.MinValue with get, set
    member val HiddenAt = nullDateTimeOffset with get, set

    interface IFavorite with
        member this.HiddenAt
            with get () = this.HiddenAt
             and set value = this.HiddenAt <- value

        member this.FavoritedAt = this.FavoritedAt

    interface IPost with
        member _.Platform = DeviantArt
        member this.Url = this.LinkUrl
        member this.DisplayTitle = this.Title |> orString ""
        member this.Id = $"{this.Id}"
        member this.InternalUrl = this.LinkUrl
        member this.ExternalUrl = this.LinkUrl
        member this.PostedAt = this.Timestamp
        member this.ProfileUrl = $"https://www.deviantart.com/{Uri.EscapeDataString(this.Username)}"
        member this.Thumbnails = [
            for url in this.ThumbnailUrls do {
                new IPostThumbnail with
                    member _.AltText = ""
                    member _.Url = url
            }
        ]
        member this.Usericon = this.Usericon
        member this.Username = this.Username

File: Pandacap.Database/DeviantArtFavorite.cs
using Pandacap.UI.Badges;
using Pandacap.UI.Elements;

namespace Pandacap.Database
{
    public class DeviantArtFavorite : IFavorite
    {
        public Guid Id { get; set; }

        public Guid CreatedBy { get; set; }

        public string Username { get; set; } = "";

        public string? Usericon { get; set; }

        public DateTimeOffset Timestamp { get; set; }

        public string? Title { get; set; }

        public string? Content { get; set; }

        public string LinkUrl { get; set; } = "";

        public List<string> ThumbnailUrls { get; set; } = [];

        public DateTimeOffset FavoritedAt { get; set; }

        public DateTimeOffset? HiddenAt { get; set; }

        Badge IPost.Badge => Badges.DeviantArt.WithHostFromUriString(LinkUrl);

        string IPost.DisplayTitle => Title ?? "";

        string IPost.Id => $"{Id}";

        string IPost.InternalUrl => LinkUrl;

        string IPost.ExternalUrl => LinkUrl;

        DateTimeOffset IPost.PostedAt => Timestamp;

        string? IPost.ProfileUrl => $"https://www.deviantart.com/{Uri.EscapeDataString(Username)}";

        IEnumerable<IPostThumbnail> IPost.Thumbnails => ThumbnailUrls.Select(url => new PostThumbnail(url));

        private record PostThumbnail(string Url) : IPostThumbnail
        {
            public string AltText => "";
        }
    }
}
